dnsmasq: update to 2.87test8+

aparcar opened the pull request:
With the recent switch to firewall4 which uses nftables, we need dnsmasq
support for nftables as well.

Signed-off-by: Paul Spooren <mail@aparcar.org>

ldir-EDB0 committed:
dnsmasq: remove check for existing dhcp server

Remove check for existing dhcp server on interface.  This check could
already be overridden by using 'force' parameter so this change
forces 'force'.

Signed-off-by: Kevin Darbyshire-Bryant <kevin@darbyshire-bryant.me.uk>

ldir-EDB0 committed:
dnsmasq: use hostsdir instead of addn-hosts

dnsmasq will dynamically scan changes/additions to hosts in hostsdir whereas
addn-hosts needs a SIGHUP.

Take advantage of this automatic updating in preparation for a dhcpd
update take only signals
addn-hosts

-       xappend "--addn-hosts=$(dirname $HOSTFILE)"
+       xappend "--hostsdir=$(dirname $HOSTFILE)"

Signed-off-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>

ldir-EDB0 committed:
dnsmasq: source RFC6761 file first

Signed-off-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>

rsalvaterra committed:
dnsmasq: add a new "bind" configuration option

The nonwildcard option only allows to choose between the default behaviour (bind
to the wildcard address only, which unfeasible when running multiple instances
of dnsmasq) and --bind-dynamic (bind to any existing/new interfaces which aren't
explicitly configured otherwise). Add a new "bind" option, which can take the
value of "dynamic" or "interfaces", allowing the user to additionally choose the
--bind-interfaces behaviour (bind only to explicitly configured interfaces).

The dnsmasq man page [1] contains more detailed information about these
arguments. Finally, the nonwildcard option is kept for backwards compatibility.

[1] https://thekelleys.org.uk/dnsmasq/docs/dnsmasq-man.html

Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>

ldir-EDB0 committed:
dnsmasq: bump to 2.87test8

Signed-off-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>

ldir-EDB0 committed:
dnsmasq: enable dnsmasq's new nftables support

Add build option for nftables sets.

Signed-off-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>
[Fix typos]
Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>

rsalvaterra committed:
dnsmasq: add a new "dnsfilter" configuration option

Add UCI support for record filtering in dnsmasq replies. The new "dnsfilter"
option can take either "A" or "AAAA" as possible values, enabling filtering of
IPv4 or IPv6 addresses, respectively.

Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>

hnyman commented:
cc @ldir who has had dnsmasq stuff in his staging repo for some time already...

https://git.openwrt.org/?p=openwrt/staging/ldir.git;a=shortlog;h=refs/heads/mine

dave14305 commented:
To get nftables set support, Makefile will need to be updated to build dnsmasq with HAVE_NFTSET option.

See https://git.openwrt.org/?p=openwrt/staging/ldir.git;a=commitdiff;h=cdaa993f4cc9c65792a923c4903efafd3b8e4eb0

aparcar commented:
I'm closing this in favor of the work happening in [Kevins staging dir](https://git.openwrt.org/?p=openwrt/staging/ldir.git;a=shortlog;h=refs/heads/mine)

ldir-EDB0 committed:
dnsmasq: include latest patches

Signed-off-by: Kevin Darbyshire-Bryant <ldir@darbyshire-bryant.me.uk>

dave14305 commented:
Changes are also needed to dnsmasq.init to add the relevant nftset options to the dnsmasq.conf file, using the existing ipset uci config:
```
nftset=/domain1.com/domain2.com/4#inet#fw4#setname4
nftset=/domain1.com/domain2.com/6#inet#fw4#setname6
```

rsalvaterra committed:
dnsmasq: support populating nftsets in addition to ipsets

Tell dnsmasq to populate nftsets instead of ipsets, if firewall4 is present in
the system. Keep the same configuration syntax in /etc/config/dhcp, for
compatibility purposes.

Huge thanks to Jo-Philipp Wich for basically writing the function.

Signed-off-by: Jo-Philipp Wich <jo@mein.io>
Signed-off-by: Rui Salvaterra <rsalvaterra@gmail.com>

dave14305 commented:
Since 22.03 rc1 was tagged today, what are the prospects for dnsmasq being ready before 22.03 release? I don’t think Simon Kelley will give us an official 2.87 release anytime soon.

mkresin commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -170,7 +170,26 @@ append_address() {
+		done
+		IFS=$' \t\n'
+
+		xappend "--nftset=$domainlist/$nftsets"
```
Please add a check if `nftset` is supported by dnsmasq, similar to the `ipset` check. Otherwise dnsmasq should refuse to start.

Kevin has already such a check in hist dnsmasq bump [in his staging tree](https://git.openwrt.org/?p=openwrt/staging/ldir.git;a=shortlog;h=refs/heads/mine).

mkresin commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -170,7 +170,26 @@ append_address() {
 }
 
 append_ipset() {
```
This change will only cover the "old way" of defining ipsets.

We have two option to define ipsets within the dnsmasq section:

```
config dnsmasq
	list   ipset '/openwrt.org/openwrt_ipv4,openwrt_ipv6'
```

And since acb336235cc640980ea549655b0702daf9e628f8 in addition

```
config ipset
            list name 'openwrt_ipv4'
            list name 'openwrt_ipv6'
            list domain 'openwrt.org'
```

May I propose something else:

1. migrate the old style to the new one
2. drop support for the old style
3. add the changes to `dnsmasq_ipset_add` instead

Based on Kevins changes I've already written the code doing so. Feel free to pick the changes from [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git).

mkresin commented on package/network/services/dnsmasq/Makefile:
```diff
@@ -110,6 +112,9 @@ define Package/dnsmasq-full/config
 	config PACKAGE_dnsmasq_full_ipset
 		bool "Build with IPset support."
 		default y
```
Since `nftables` is the new default, it might be a good idea to disable IPset support by default.

mkresin commented on package/network/services/dnsmasq/files/dhcp.conf:
```diff
@@ -15,7 +15,7 @@ config dnsmasq
 	option resolvfile	'/tmp/resolv.conf.d/resolv.conf.auto'
 	#list server		'/mycompany.local/1.2.3.4'
-	option nonwildcard	1 # bind to & keep track of interfaces
+	option bind		'dynamic' # bind to & keep track of interfaces
```
Shouldn't a migration script be added to update existing configs similar to package/network/services/dnsmasq/files/50-dnsmasq-migrate-resolv-conf-auto.sh?

mkresin commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -170,7 +170,26 @@ append_address() {
 
 append_ipset() {
-	xappend "--ipset=$1"
+	if [ -x /sbin/fw4 ]; then
```
The check for fw4 shouldn't be required. We either have iptables or nftables. If nftables is present, dnsmasq fails silently to add the ipsets. Means, this function can append the ntfset and ipset parameter at the same time.

You can find such an implementation in my staging tree.

mkresin commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -170,7 +170,26 @@ append_address() {
+
+		IFS=,
+		for set in $setlist; do
+			local family=$(nft -t list set inet fw4 "$set" 2>&1 | sed -nre 's#^\t\ttype .*\bipv([46])_addr\b.*$#\1#p')
```
No need to take care of the sets ip family type. dnsmasq will try to add the IP addresses to all sets, regardless of the IP version defined for the set. Adding an IPv6 to an IPv4 set and vice versa will silently fail. In the end, the sets are populated correctly.

According to the dnsmasq man page, it's the same behaviour as with ipsets.

Once more, you can find such an implementation in my staging tree.

PolynomialDivision commented:
Meanwhile dnsmasq-2.87 is released. ;)

dhewg commented:
See also #10820

981213 commented:
dnsmasq is updated to 2.87 already. Closing this PR :)
